### DISCUSSION TITLE
Solving with custom constraints

### DISCUSSION THREAD

**Discussion by ajinvm** [google/or-tools#4086] [Routing (and legacy CP) questions]
I am trying to solve a problem with Python - Ortools with timewindow,capacity constraints.

I have an unique requirement where 

1. i have a set of female and male employees as waypoints,
2. there are two types of trips drop and pickup
3. in both trips female employees should not be allowed to travel with the driver alone
4. So in pickup the first employee should not be female, if first employee is female then add a escort/guard so only it can pickup Capcity-1 number of employees
5. For drop last employee should not be female. if female it should add a guard/escort

How can i acheive this?
Please someone help

Thanks

**Comment by Mizux** [comment]
I would try:

* Create a capacity dimension to track male employee number
* Create a capacity dimension to track female employee number
* For each node, add a constraint that `CapacityMaleDimension.CumulVar(index) + CapacityFemaleDimension.CumluVar(index) < VehicleMaxCapacity` (since we have split the capacity in two dimensions you must have a constraint on the sum)
* For each node, add a constraint: `2 * CapacityMaleDimension.CumulVar(location_index) + CapacityFemaleDimension.CumulVar(location_index) >= 2` (one male employee will be counted as 2 so he can travel alone)

**Reply by ajinvm** [reply]
Thanks for the reply.
But i think i did not ellaborate the problem well.

The goal of the problem is that the female employees can travel with male employees. But they are not allowed to travel with driver only at any point of the trip.

If its pickup and first employee is male -> then other employees does not matter for this vehicle , because the first male employee acts as a guard untill all employees are picked and dropped at office.

for drop -> if last employee is male then he acts as a guard, untill all female employees are dropped.

Is there a way to solve this with constraints?

**Reply by Mizux** [reply]
that's what I wrote, except I also allow the vehicle to travel with two female employee or above (no need guard when not alone).

If you always want at least a male employee you can use:
`FemaleCapacityDim.CumulVar(index)  < FemaleCapacityDim.CumulVar(index)  + MaleCapacityDim.CumulVar(Index)`

**Comment by ajinvm** [comment]
Ok thanks , i will try this and let u know

**Reply by lperron** [reply]
Or forbid any initial route from depot to female pickup or from female delivery to depot.

**Reply by ajinvm** [reply]
Hi, I need some help on this

How can i add a soft constraint to 
1. Pick a male employee as the first waypoint during pickup trip
2. Pick a male employee as last waypoint during drop trips?

These should be soft constraints.

Thanks

**Comment by lperron** [comment]
Penalize arcs depot -> male or male -> depot?
